### PR TITLE
open-mesh: make qt optional

### PR DESCRIPTION
**Pull request by autosquid**
qt dependency is only optional according to build instructions from official doc.

### CODE REVIEW

File: Library/Formula/open-mesh.rb [hunk 1]
```diff
@@ -16,11 +16,18 @@ class OpenMesh < Formula
...
-  depends_on "qt"
+  option "with-qt", "Install Qt if you want to use the included applications"
+  depends_on "qt" => :optional
 
   def install
     mkdir "build" do
-      system "cmake", "..", *std_cmake_args
+      args = std_cmake_args
+      if build.with? "qt"
+        arg << "-DBUILD_APPS=ON"
```
Comment (DomT4) [inline Library/Formula/open-mesh.rb]: Should be `args` on both these lines.

File: Library/Formula/open-mesh.rb [hunk 2]
```diff
@@ -16,11 +16,18 @@ class OpenMesh < Formula
   head "http://openmesh.org/svnrepo/OpenMesh/trunk/", :using => :svn
 
   depends_on "cmake" => :build
-  depends_on "qt"
+  option "with-qt", "Install Qt if you want to use the included applications"
+  depends_on "qt" => :optional
```
Comment (DomT4) [inline Library/Formula/open-mesh.rb]: Let's just keep `depends_on "qt" => :optional` and ditch the `option` line.

File: Library/Formula/open-mesh.rb (hunk 1 above)
Comment (autosquid) [inline Library/Formula/open-mesh.rb]: thanks for pointing out.

File: Library/Formula/open-mesh.rb (hunk 2 above)
Comment (autosquid) [inline Library/Formula/open-mesh.rb]: ditched in new push. Do you know how to add option hints to `depends_on "qt"`?

**Comment by mikemcquaid** [comment]
Build failing: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/29171/version=yosemite/testReport/junit/brew-test-bot/yosemite/install_open_mesh/

File: Library/Formula/open-mesh.rb (hunk 2 above)
Comment (DomT4) [inline Library/Formula/open-mesh.rb]:
```
--with-libquicktime
	Build with libquicktime support
```

Similar to that.

**Comment by autosquid** [comment]
@mikemcquaid  It seems there is something wrong in the `CMakeLists.txt` provided by open-mesh. With `BUILD_APP` off, the `fix_bundle.cmake` file is not generated.  

A hotfix to this is  adding the following line to `line 63`:
`    configure_file ("${CMAKE_SOURCE_DIR}/cmake/fixbundle.cmake.in" "${CMAKE_BINARY_DIR}/fixbundle.cmake" @ONLY IMMEDIATE)
`

I do not know Ruby at all, and its my first try on a homebrew formular, I tried but could not figure out how the patch thing works.  Could you help with that?

File: Library/Formula/open-mesh.rb [hunk 3]
```diff
@@ -16,11 +16,17 @@ class OpenMesh < Formula
   head "http://openmesh.org/svnrepo/OpenMesh/trunk/", :using => :svn
 
   depends_on "cmake" => :build
-  depends_on "qt"
+  depends_on "qt" => :optional
```
Comment (guoxiao) [inline Library/Formula/open-mesh.rb]: Should this be `depends_on "qt" if build.with? "qt"` ?

Comment (DomT4) [inline Library/Formula/open-mesh.rb]: Nope :wink:

Comment (autosquid) [inline Library/Formula/open-mesh.rb]: To be concrete, I do not know when and where the `patch` command will be called. Is `patch` applied on untarred files? under the root dir?

Comment (guoxiao) [inline Library/Formula/open-mesh.rb]: @DomT4 So what's the recommended usage of `:optional` and `if build.with?` ? I'm often confused when using things like this.

Comment (DomT4) [inline Library/Formula/open-mesh.rb]: > To be concrete, I do not know when and where the patch command will be called. Is patch applied on untarred files? under the root dir?

The tarball is untarred into the `buildpath`, and then any patching/inreplaces are applied in the `buildpath` before anything else begins.

> So what's the recommended usage of :optional

Dependencies should always (Ideally) be mandatory, which is just:

```ruby
depends_on "openssl"
```

Recommended (Built with by default, required for bottled installations, available for the user to manually reject) -

```ruby
depends_on "openssl" => :recommended
```

Option dependencies aren't built with by default, not required for either bottled or from-source installs, but available to the user to manually ask `brew` to install with - 

```ruby
depends_on "openssl" => :optional
```

The only real reason dependencies should be `if build.with?` is if they only apply in certain conditions, aren't triggered by the dependency, and asking the user to specify all the deps is meh:

```
depends_on "openssl" => :optional
depends_on "jpeg" if build.with? "openssl"
```

In this case `jpeg` isn't a dependency of `openssl` so it wouldn't be called, but if for some reason the formula requires you to have `jpeg` installed and available to build with `openssl` support the extra dependency needs to be declared as above.

`if build.with?` and `if build.without?` are largely intended for use inside the `def install` block, to declare which arguments should be applied in which situations.

```ruby
if build.with? "openssl"
  args << "--with-openssl=yup"
else
  args << "--without-openssl"
end
```

Comment (DomT4) [inline Library/Formula/open-mesh.rb]: Sorry, typo there.

> Dependencies should always (Ideally) be mandatory, which is just:

Should have been

> Dependencies should always (Ideally) be mandatory/optional/recommended, which is just:

Fixed now.

Comment (guoxiao) [inline Library/Formula/open-mesh.rb]: @DomT4 
According to what you have just replied, am I to understand that `depends_on "openssl" => :optional` is equivalent to the following two lines:
```
option "with-openssl"
depends_on "openssl" if build.with "openssl"
```

Comment (DomT4) [inline Library/Formula/open-mesh.rb]: Yes, essentially.

If you do `brew info` on something that has an optional dep you'll see there's an automatic `:optional` option. [Caudec](https://github.com/Homebrew/homebrew/blob/master/Library/Formula/caudec.rb) has about 16 optional deps but no `option` lines, and unless there's an `option` line given to override the default we print:

```
==> Options
--with-aacgain
	Build with aacgain support
--with-alac
	Build with alac support
--with-cksfv
	Build with cksfv support
--with-coreutils
	Build with coreutils support
--with-eye-d3
	Build with eye-d3 support
--with-ffmpeg
	Build with ffmpeg support
--with-flac
	Build with flac support
--with-flake
	Build with flake support
```

And so on inside `brew info caudec`.

File: Library/Formula/open-mesh.rb [hunk 4]
```diff
@@ -1,31 +1,42 @@
...
-  head "http://openmesh.org/svnrepo/OpenMesh/trunk/", :using => :svn
+    depends_on "cmake" => :build
+    depends_on "qt" => :optional
 
-  depends_on "cmake" => :build
-  depends_on "qt"
+    patch do
+        url "https://gist.githubusercontent.com/autosquid/41741a5ff42ddb4d104c/raw/6c2b49b3b23bdbafbed053dc8dc54d3407cdfba1/CMakeLists.patch"
+        sha256 "cd7f8e58dce947f1680fc0cb10bacf208fc52b384025d1598eef31b9c6840528"
+    end
```
Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

Comment (autosquid) [inline Library/Formula/open-mesh.rb]: For open-mesh, it seems patches can only be submitted by mailling:  http://openmesh.org/participating/

I submitted the patch, but it is currently being held until the list moderator can review it for approval.

I commented why the patch is need in the formula file. Hoping I am doing right.

**Comment by autosquid** [comment]
It seems only patches by mail: http://openmesh.org/participating/
I have submitted the patch, but it is currently being held until the list moderator can review it for approval.

I commented in the formula explaining why the patch is needed.

Hope I am doing right.

File: Library/Formula/open-mesh.rb [hunk 5]
```diff
@@ -1,31 +1,46 @@
...
-      system "cmake", "..", *std_cmake_args
-      system "make install"
+    desc "Generic data structure to represent and manipulate polygonal meshes"
+    homepage "http://openmesh.org"
+
+    url "http://www.openmesh.org/media/Releases/3.3/OpenMesh-3.3.tar.gz"
+    sha1 "673677702a27a929c9124f8d4ea2b188a0500c50"
+
+    bottle do
+        cellar :any
```
Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Don't reindent the file, please, keep it all at two-character indentation.

File: Library/Formula/open-mesh.rb [hunk 6]
```diff
@@ -1,31 +1,46 @@
...
+            else
+                args << "-DBUILD_APPS=OFF"
+            end
+            system "cmake", "..", *args
+            system "make install"
+        end
+    end
+
+    test do
+        system "#{bin}/mconvert", "-help"
```
Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Test is failing: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/29242/version=yosemite/testReport/junit/brew-test-bot/yosemite/test_open_mesh/

**Comment by autosquid** [comment]
why this branch is still pending.. 2 days has passed..

**Comment by DomT4** [comment]
@BrewTestBot test this please

**Comment by mikemcquaid** [comment]
@autosquid because we're doing this in our spare time and aren't paid.

**Comment by mikemcquaid** [comment]
@autosquid Please update this to use a `sha256` hash (from `sha1`), thanks!

**Comment by autosquid** [comment]
@mikemcquaid  oh, I used to think there was a big server, and tests are run automatically - -.   the `sha1` things comes from the original formula, but I have changed it to `sha256`

**Comment by autosquid** [comment]
so, what should I do next?

File: Library/Formula/open-mesh.rb [hunk 7]
```diff
@@ -16,16 +16,70 @@ class OpenMesh < Formula
...
-  depends_on "qt"
+  depends_on "qt" => :optional
+
+  # For 3.3 version, when BUILD_APPS=OFF, the fixbundle.cmake will not be
+  # generated for a lacking cmake command `configure_file`. I patched the
+  # cmakelists.txt a little by adding the missing line.
+  patch do
+    url "https://gist.githubusercontent.com/autosquid/41741a5ff42ddb4d104c/raw/6c2b49b3b23bdbafbed053dc8dc54d3407cdfba1/CMakeLists.patch"
+    sha256 "cd7f8e58dce947f1680fc0cb10bacf208fc52b384025d1598eef31b9c6840528"
+  end
```
Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Can you please submit this patch to the upstream developers of this project and add a link to the upstream patch submission and explanation of why the patch is needed in a comment in the formula file. Thanks!

Comment (autosquid) [inline Library/Formula/open-mesh.rb]: already submitted but with no reply.  Explanations are already included, PLZ check

Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Please add a link to the upstream patch submission.

Comment (autosquid) [inline Library/Formula/open-mesh.rb]: for non-members, patches are submitted only via mail. so no such link stuff.  see [http://openmesh.org/participating/]

Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Can you note when and where you submitted the comment? Thanks!

Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: e.g. the email address and date

**Comment by DomT4** [comment]
Needs rebasing on master here.

**Comment by autosquid** [comment]
I have rebased (and squashed) my commits based on the most recent master branch. Plz check.

File: Library/Formula/open-mesh.rb [hunk 8]
```diff
@@ -1,29 +1,82 @@
...
   homepage "http://openmesh.org"
-  url "http://www.openmesh.org/media/Releases/3.3/OpenMesh-3.3.tar.gz"
-  sha256 "1fbc458e6f7d43d24a0e13005c094d15f08cc2f734a943d6c5417f0e6d9ad70e"
-
-  bottle do
-    cellar :any
-    sha256 "e4bc6932d6a585fa96b9cc8f09c3037d852dae79072a8feea47612ba48598212" => :yosemite
-    sha256 "7806df7d4f499a801350087017944a79365336aec3b266bb5bee2ef17ffa017a" => :mavericks
-    sha256 "0ca60ff704a7d442bdd432d9a63f58d2e453d787648f3831bcaa27ac2f5a1114" => :mountain_lion
-  end
```
Comment (mikemcquaid) [inline Library/Formula/open-mesh.rb]: Please leave the bottle block as-is and it'll be updated when we pull, thanks.

Comment (autosquid) [inline Library/Formula/open-mesh.rb]: ok, add that back. plz. check.

**Comment by autosquid** [comment]
Is there anything  I can do (or suggestions) to make this RP. better? @DomT4  @mikemcquaid

**Comment by DomT4** [comment]
I think it's more or less fine but I'll let Mike look over it to check there's nothing he wants tweaked.

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

For future reference the preferred commit message format for simple version updates is `open-mesh 4.0 (and optional qt)`.

**Pull request closed without merging**